### PR TITLE
Update library to use new architecture.

### PR DESCRIPTION
**Pull request by jhtimmins**
This adds breaking changes to the library and would necessitate a major version update.
Among other things, this branch:
- Changes the name.
- Adopts an OOP architecture.
- Adds caching.
- Adds the country_name, ip_address, latitude, and longitude attributes to the response object.
- Updates documentation to match current IPinfo specs.
- Modifies the request headers.
- Extends tests.

### CODE REVIEW

**Comment by coderholic** [comment]
I know the repl style examples with `>>>` are popular for python, but is that also the case for ruby? I looked at a couple of gems and they didn't seem to.

**Comment by skatkov** [comment]
Sorry to cheap in, just wanted to give some feedback before you merge this.

While there is on PEP-8 alternative in Ruby, there is still some community style guides that most of teams that use Ruby respect. You can get some taste of it here
https://github.com/rubocop-hq/ruby-style-guide

Code I see here would be normal for Java, but it makes me cringe as rubyist. Namely:

- prepending *get_* and *set_* to method names  -- https://github.com/rubocop-hq/ruby-style-guide#accessor_mutator_method_names
- CamelCase for methods -- https://github.com/rubocop-hq/ruby-style-guide#snake-case-symbols-methods-vars
- Interface class for cache seems like a leaky abstraction -- it brings no value to code, used once and gets immediately redefined. I would suggest dropping it, but if you plan on leaving it -- at least make it useful.

```
module IPinfo
  class CacheInterface
    class InterfaceNotImplemented < StandardError; end
    def get(key)
        raise InterfaceNotImplemented
    end
    def set(key, value)
        raise InterfaceNotImplemented
    end
    def contains(key)
        raise InterfaceNotImplemented
    end
  end
end
```

But again, API for this interface just feels like a person is writing Java :)

File: lib/ipinfo/adapter.rb [hunk 1]
```diff
@@ -33,10 +33,15 @@ def connection(adapter)
     end
 
     def default_headers
-        {
-          'User-Agent' => "ipinfo-ruby/#{::IpinfoIo::VERSION}",
+        headers = {
+          'User-Agent' => "IPinfoClient/Ruby/1.0",
```
Comment (skatkov) [inline lib/ipinfo/adapter.rb]: It seems like a duplication -- version number is stored in version.rb file, but your hardcoding it here. I think people will forget to increment it here.

File: lib/ipinfo/adapter.rb [hunk 2]
```diff
@@ -33,10 +33,15 @@ def connection(adapter)
...
           'Accept' => 'application/json'
         }
+        if token
+          headers['Authorization'] = "Bearer #{CGI::escape(token)}"
+        end
+        puts headers
```
Comment (skatkov) [inline lib/ipinfo/adapter.rb]: Would be great to not have a debugging code in a library, nobody likes libraries spitting out tons of debugging info without ability to turn it off.

File: lib/ipinfo/cache/default_cache.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+    def get(key)
+      @cache[key]
+    end
+
+    def set(key, value)
+      @cache[key] = value
+    end
+
+    def contains(key)
```
Comment (skatkov) [inline lib/ipinfo/cache/default_cache.rb]: this returns boolean, so method should end with question mark. e.g. contains? 

https://github.com/rubocop-hq/ruby-style-guide#bool-methods-prefix

File: lib/ipinfo/adapter.rb (hunk 1 above)
Comment (jhtimmins) [reply lib/ipinfo/adapter.rb]: In this case, the version is actually intended to be the API version, not the library version.

**Comment by jhtimmins** [comment]
Thanks so much @skatkov, this is super helpful and I really appreciate the feedback. In this case I've opted to keep the interface, but that's mostly bc that's a design we've been using across libraries. Thanks again!

File: lib/ipinfo.rb [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+require 'json'
+
+module IPinfo
+  DEFAULT_CACHE_MAXSIZE = 4096
+  DEFAULT_CACHE_TTL = 60 * 60 * 24
+  DEFAULT_COUNTRY_FILE = File.join(File.dirname(__FILE__), 'ipinfo/countries.json')
+  RATE_LIMIT_MESSAGE = "To increase your limits, please review our paid plans at https://ipinfo.io/pricing"
+
+  class << self
+    def handler(access_token=nil, settings={})
```
[the remark this answers is not in the document]
Comment (jhtimmins) [reply lib/ipinfo.rb]: Thanks very much!

**Pull request merged**